plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -627,92 +663,96 @@ type ServiceAffinity struct {
 	labels        []string
 }
 
-func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister algorithm.ServiceLister, nodeInfo NodeInfo, labels []string) algorithm.FitPredicate {
+func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister algorithm.ServiceLister, nodeInfo NodeInfo, labels []string) (algorithm.FitPredicate, func(pm *predicateMetadata)) {
 	affinity := &ServiceAffinity{
 		podLister:     podLister,
 		serviceLister: serviceLister,
 		nodeInfo:      nodeInfo,
 		labels:        labels,
 	}
-	return affinity.CheckServiceAffinity
+
+	return affinity.CheckServiceAffinity, func(pm *predicateMetadata) {
+		var err error
+		// We can't know at the start what labels we are selecting against.
+		pm.servicePods[""], err = podLister.List(createSelectorFromLabels(nil))
```

Comment:
This isn't what I had in mind.
Basically, we should extend predicateMetadata to contain the pod (or just its selector) and use this selector here.

As for podLister, we should use the mechanism that is already present for prioityMetadata (we should just duplicate it for predicates):
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go#L64

Basically, we should add a similar method for predicatemetadata:
"RegisterPredicateMetadataProducerFactory"
and then using the args from here we would be able to make the "podLister" be a filed of "PredicateMetadataFactory".